### PR TITLE
[web] Paragraph.getBoxesForRange uses LineMetrics

### PR DESCRIPTION
**Pull request by mdebbar**
Fixes https://github.com/flutter/flutter/issues/39826

### CODE REVIEW

File: lib/web_ui/lib/src/engine/text/paragraph.dart [hunk 1]
```diff
@@ -363,13 +365,59 @@ class EngineParagraph implements ui.Paragraph {
...
+
+    final double top = line.lineNumber * _lineHeight;
+
+    //               |<------------------ line.width ------------------>|
+    // |-------------|------------------|-------------|-----------------|
+    // |<-line.left->|<-widthBeforeBox->|<-box width->|<-widthAfterBox->|
+    // |-------------|------------------|-------------|-----------------|
+    //
+    //                                   ^^^^^^^^^^^^^
+    //                          This is the box we want to return.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text/paragraph.dart]: Love the ASCII art docs :+1:

File: lib/web_ui/test/paragraph_test.dart [hunk 2]
```diff
@@ -426,6 +426,209 @@ void main() async {
     expect(paragraph.getBoxesForRange(0, 0), isEmpty);
   });
 
+  testEachMeasurement('getBoxesForRange multi-line', () {
```
Comment (yjbanov) [inline lib/web_ui/test/paragraph_test.dart]: Is it worth adding a test for a `ParagraphStyle` with `maxLines` such that the `start`/`end` range fall outside the max line bounds?

Comment (mdebbar) [reply lib/web_ui/test/paragraph_test.dart]: That's a good corner case to test. I'd like to merge this PR while the tree is green though. So I'll add the `maxLines` test in a future PR.
